

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Quick start &mdash; igphyml 1.0.7 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
    <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Release Notes" href="../NEWS.html" />
    <link rel="prev" title="Installation" href="../install.html" /> 

  
  <script src="../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../index.html" class="icon icon-home"> igphyml
          

          
            
            <img src="../_static/logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Getting Started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
</ul>
<p class="caption"><span class="caption-text">Using IgPhyML</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Quick start</a></li>
<li class="toctree-l1"><a class="reference internal" href="#processing-change-o-data-sets">Processing Change-O data sets</a></li>
<li class="toctree-l1"><a class="reference internal" href="#building-b-cell-lineage-trees">Building B cell lineage trees</a></li>
<li class="toctree-l1"><a class="reference internal" href="#phylogenetic-model-parameter-analysis">Phylogenetic model parameter analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="#optimizing-performance">Optimizing performance</a></li>
</ul>
<p class="caption"><span class="caption-text">About</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../NEWS.html">Release Notes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../info.html">Contact</a></li>
<li class="toctree-l1"><a class="reference internal" href="../info.html#citation">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../info.html#license">License</a></li>
</ul>
<p class="caption"><span class="caption-text">Related Projects</span></p>
<ul>
<li class="toctree-l1"><a class="reference external" href="http://immcantation.readthedocs.io">Immcantation Portal</a></li>
<li class="toctree-l1"><a class="reference external" href="http://changeo.readthedocs.io">Change-O</a></li>
<li class="toctree-l1"><a class="reference external" href="http://alakazam.readthedocs.io">Alakazam</a></li>
<li class="toctree-l1"><a class="reference external" href="http://presto.readthedocs.io">pRESTO</a></li>
<li class="toctree-l1"><a class="reference external" href="http://shazam.readthedocs.io">SHazaM</a></li>
<li class="toctree-l1"><a class="reference external" href="http://tigger.readthedocs.io">TIgGER</a></li>
<li class="toctree-l1"><a class="reference external" href="http://rdi.readthedocs.io">RDI</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">igphyml</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
      <li>Quick start</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/usage/quickstart.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="quick-start">
<h1>Quick start<a class="headerlink" href="#quick-start" title="Permalink to this headline">¶</a></h1>
<p>These commands should work as a first pass on many reasonably sized
datasets, but if you really want to understand what’s going on or make
sure what you’re doing makes sense, please check out the rest of the
manual.</p>
<p><strong>Convert Change-O files into IgPhyML inputs</strong></p>
<p>Move to the <code class="docutils literal notranslate"><span class="pre">examples</span></code> subfolder and run:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">BuildTrees</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">d</span> <span class="n">example</span><span class="o">.</span><span class="n">tab</span> <span class="o">--</span><span class="n">outname</span> <span class="n">ex</span> <span class="o">--</span><span class="n">log</span> <span class="n">ex</span><span class="o">.</span><span class="n">log</span> <span class="o">--</span><span class="n">collapse</span> \
    <span class="o">--</span><span class="n">sample</span> <span class="mi">3000</span> <span class="o">--</span><span class="n">minseq</span> <span class="mi">2</span>
</pre></div>
</div>
<p><strong>Build lineage trees using the GY94 model</strong> (fast, doesn’t correct
for hotspots):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">igphyml</span> <span class="o">--</span><span class="n">repfile</span> <span class="n">ex_lineages</span><span class="o">.</span><span class="n">tsv</span> <span class="o">-</span><span class="n">m</span> <span class="n">GY</span>
</pre></div>
</div>
<p><strong>Build lineage trees using the HLP19 model</strong> (slower, corrects for
WRC/GYW hotspots):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">igphyml</span> <span class="o">--</span><span class="n">repfile</span> <span class="n">ex_lineages</span><span class="o">.</span><span class="n">tsv</span> <span class="o">-</span><span class="n">m</span> <span class="n">HLP</span>
</pre></div>
</div>
<p>Both of these can be parallelized by adding
<code class="docutils literal notranslate"><span class="pre">--threads</span> <span class="pre">&lt;thread</span> <span class="pre">count&gt;</span></code> option. Trees files are listed as
<code class="docutils literal notranslate"><span class="pre">ex/&lt;clone</span> <span class="pre">id&gt;.fasta_igphyml_tree.txt</span></code>, and can be viewed with most
tree viewers (I recommend
<a class="reference external" href="http://tree.bio.ed.ac.uk/software/figtree/">FigTree</a>). Parameter
estimates are in <code class="docutils literal notranslate"><span class="pre">ex_lineages.tsv_igphyml_stats.txt</span></code>. Subsampling using
the <code class="docutils literal notranslate"><span class="pre">--sample</span></code> option in BuildTrees isn’t strictly necessary, but
IgPhyML will run slowly when applied to large datasets.</p>
</div>
<div class="section" id="processing-change-o-data-sets">
<span id="buildtrees-processing"></span><h1>Processing Change-O data sets<a class="headerlink" href="#processing-change-o-data-sets" title="Permalink to this headline">¶</a></h1>
<p>The process begins with a Change-O formatted data file, in
which each sequence has been clustered into a clonal group,
which has subsequently had its unmutated V and J sequence predicted germlines.
The following column names are required in the input file: fields: <code class="docutils literal notranslate"><span class="pre">SEQUENCE_ID</span></code>,
<code class="docutils literal notranslate"><span class="pre">SEQUENCE_INPUT</span></code>, <code class="docutils literal notranslate"><span class="pre">SEQUENCE_IMGT</span></code>, <code class="docutils literal notranslate"><span class="pre">GERMLINE_IMGT_D_MASK</span></code>,
<code class="docutils literal notranslate"><span class="pre">V_CALL</span></code>, <code class="docutils literal notranslate"><span class="pre">J_CALL</span></code>, and <code class="docutils literal notranslate"><span class="pre">CLONE</span></code>. <code class="docutils literal notranslate"><span class="pre">FUNCTIONAL</span></code> is recommended.</p>
<p>Use BuildTrees.py to break this file into separate sequence
alignment files that can be used with IgPhyML. This program will:</p>
<ol class="arabic simple">
<li>Filter out nonfunctional sequences.</li>
<li>Mask codons split by insertions.</li>
<li>Separate clonal groups into separate alignment files (aligned by IMGT site) and information files</li>
<li>Create the repertoire files for this dataset.</li>
</ol>
<p>Create IgPhyML input files from <code class="docutils literal notranslate"><span class="pre">examples/example.tab</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">cd</span> <span class="n">examples</span>
<span class="n">BuildTrees</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">d</span> <span class="n">example</span><span class="o">.</span><span class="n">tab</span> <span class="o">--</span><span class="n">outname</span> <span class="n">ex</span> <span class="o">--</span><span class="n">log</span> <span class="n">ex</span><span class="o">.</span><span class="n">log</span> <span class="o">--</span><span class="n">collapse</span>
</pre></div>
</div>
<p>This will create the directory <code class="docutils literal notranslate"><span class="pre">ex</span></code> and the file
<code class="docutils literal notranslate"><span class="pre">ex_lineages.tsv</span></code>. Each <code class="docutils literal notranslate"><span class="pre">ex/&lt;clone</span> <span class="pre">ID&gt;.fasta</span></code> contains the IMGT
mutliple sequence alignemt for a particular clone, and each
<code class="docutils literal notranslate"><span class="pre">ex/&lt;clone</span> <span class="pre">ID&gt;.part.txt</span></code> file contains information about V and J
germline assignments, as well as IMGT unique numbering for each site.
The file <code class="docutils literal notranslate"><span class="pre">ex.log</span></code> will contain information about whether or not each
sequence was included in the analysis. The file <code class="docutils literal notranslate"><span class="pre">ex_lineages.tsv</span></code> is
the direct input to IgPhyML. Each line represents a clone and shows
the multiple sequence alignment, starting tree topology (N if
ignored), germline sequence ID in alignment file, and partition file
(N if ignored). These repertoire files start with the number of
lineages in the repertoire, and lineages are arranged from most to
least number of sequences. Here, the <code class="docutils literal notranslate"><span class="pre">--collapse</span></code> flag is used to
collapse identical sequences. This is highly recommended because
identical sequences slow down calculations without actually affecting
likelihood values in IgPhyML.</p>
<p id="buildtrees-subsampling"><strong>Subsampling Change-O datasets</strong></p>
<p>IgPhyML runs slowly with more than a few thousand sequences. You can
subsample your dataset using the <code class="docutils literal notranslate"><span class="pre">--sample</span></code> and <code class="docutils literal notranslate"><span class="pre">--minseq</span></code> options,
which will subsample your dataset to the specified depth and then remove
all clones below the specified size cutoff:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">BuildTrees</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">d</span> <span class="n">example</span><span class="o">.</span><span class="n">tab</span> <span class="o">--</span><span class="n">outname</span> <span class="n">ex</span> <span class="o">--</span><span class="n">log</span> <span class="n">ex</span><span class="o">.</span><span class="n">log</span> <span class="o">--</span><span class="n">collapse</span> <span class="o">--</span><span class="n">sample</span> <span class="mi">5</span> <span class="o">--</span><span class="n">minseq</span> <span class="mi">2</span>
</pre></div>
</div>
<p><strong>Removing the CDR3 region</strong></p>
<p>If you plan to analyze model parameters to study things such as SHM and
selection (see :ref: <cite>below &lt;igphyml-parameters&gt;</cite>), it’s important to remove
the CDR3 region to avoid known model biases in estimating <span class="math notranslate nohighlight">\(\omega\)</span>. To
do this, use <code class="docutils literal notranslate"><span class="pre">--ncdr3</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">BuildTrees</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">d</span> <span class="n">example</span><span class="o">.</span><span class="n">tab</span> <span class="o">--</span><span class="n">outname</span> <span class="n">ex</span> <span class="o">--</span><span class="n">log</span> <span class="n">ex</span><span class="o">.</span><span class="n">log</span> <span class="o">--</span><span class="n">collapse</span> <span class="o">--</span><span class="n">ncdr3</span>
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">IgPhyML requires at least three sequences in a lineage, so in
the case that there is only one observed sequence within a clone, that
sequence is duplicated. This will not affect the likelihood
calculation because these sequences will have a branch length of zero,
but it will affect metrics that take sequence frequency into account.</p>
</div>
</div>
<div class="section" id="building-b-cell-lineage-trees">
<span id="building-lineage-trees"></span><h1>Building B cell lineage trees<a class="headerlink" href="#building-b-cell-lineage-trees" title="Permalink to this headline">¶</a></h1>
<p>Before doing any further analysis, I strongly recommend estimating
intitial tree topologies using the GY94 model. This can improve
runtime for later analysis:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">igphyml</span> <span class="o">--</span><span class="n">repfile</span> <span class="n">ex_lineages</span><span class="o">.</span><span class="n">tsv</span> <span class="o">-</span><span class="n">m</span> <span class="n">GY</span> <span class="o">--</span><span class="n">outrep</span> <span class="n">ex_lineages</span><span class="o">.</span><span class="n">GY</span><span class="o">.</span><span class="n">tsv</span> <span class="o">--</span><span class="n">run_id</span> <span class="n">GY</span>
</pre></div>
</div>
<p>Here, the data files are specifed with <code class="docutils literal notranslate"><span class="pre">--repfile</span></code>. Topologies are
searched using NNI moves. To do a more thorough topology search, use
<code class="docutils literal notranslate"><span class="pre">-s</span> <span class="pre">SPR</span></code>. The flag <code class="docutils literal notranslate"><span class="pre">--outrep</span></code> will create a repertoire file that is
identical to the file specified in <code class="docutils literal notranslate"><span class="pre">--repfile</span></code> but with the resulting
GY94 topologies specified for each lineage. We can view the ML
parameter estimates for the GY94 fit in
<code class="docutils literal notranslate"><span class="pre">ex_lineages.tsv_igphyml_stats_GY.txt</span></code>, and the tree topologies for
each clone individual lineage in
<code class="docutils literal notranslate"><span class="pre">ex/&lt;clone</span> <span class="pre">id&gt;.fasta_igphyml_tree_GY.txt</span></code>. I recommend using
<a class="reference external" href="http://tree.bio.ed.ac.uk/software/figtree/">FigTree</a> to visualize
topologies.</p>
<p>To estimate ML tree topologies using the HLP19 model wth a GY94
starting topology, use:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">igphyml</span> <span class="o">--</span><span class="n">repfile</span> <span class="n">ex_lineages</span><span class="o">.</span><span class="n">GY</span><span class="o">.</span><span class="n">tsv</span> <span class="o">-</span><span class="n">m</span> <span class="n">HLP</span> <span class="o">--</span><span class="n">run_id</span> <span class="n">HLP</span> <span class="o">--</span><span class="n">threads</span> <span class="mi">2</span>
</pre></div>
</div>
<p>This will estimate a single <span class="math notranslate nohighlight">\(\omega\)</span>, <span class="math notranslate nohighlight">\(\kappa\)</span>, set of
codon frequencies (<span class="math notranslate nohighlight">\(\pi\)</span>), and WRC/GYW mutability across the
entire repertoire, and search for topologies using NNI moves. You can
see parameter estimates in
<code class="docutils literal notranslate"><span class="pre">ex_lineages.GY.tsv_igphyml_stats_HLP.txt</span></code>, and trees in
<code class="docutils literal notranslate"><span class="pre">ex/&lt;clone</span> <span class="pre">id&gt;.fasta_igphyml_tree_HLP.txt</span></code>. This command will also
parallelize the calculation across 2 threads using the <code class="docutils literal notranslate"><span class="pre">--threads</span></code>
flag.</p>
</div>
<div class="section" id="phylogenetic-model-parameter-analysis">
<span id="igphyml-parameters"></span><h1>Phylogenetic model parameter analysis<a class="headerlink" href="#phylogenetic-model-parameter-analysis" title="Permalink to this headline">¶</a></h1>
<p>The HLP19 model is the heart of IgPhyML and adjusts for features of
affinity maturation that violate the assumptions of most other
phylogenetic models. It uses four sets of parameters to characterize
the types of mutations the occurred over a lineage’s development, and
to help build the tree.</p>
<p><span class="math notranslate nohighlight">\(\omega\)</span>: Also called dN/dS, or the ratio of nonsynonymous
(amino acid replacement) and synonymous (silent) mutation rates. This
parameter generally relates to clonal selection, with totally neutral
amino acid evolution having an <span class="math notranslate nohighlight">\(\omega \approx 1\)</span>, negative
selection indicated by <span class="math notranslate nohighlight">\(\omega &lt; 1\)</span> and diversifying selection
indicated by <span class="math notranslate nohighlight">\(\omega &gt; 1\)</span>. Generally, we find a lower <span class="math notranslate nohighlight">\(\omega\)</span>
for FWRs than CDRs, presumably because FWRs are more structurally
constrained.</p>
<p><span class="math notranslate nohighlight">\(\kappa\)</span>: Ratio of transitions (within purines/pyrimidines) to
transversions (between purines/pyrimidines). For normal somatic
hypermutation this ratio is usually <span class="math notranslate nohighlight">\(\approx 2\)</span>.</p>
<p>Motif mutability (e.g. <span class="math notranslate nohighlight">\(h^{WRC}\)</span>): Mutability parameters for
specified hot- and coldspot motifs. These estimates are equivalent to
the fold-change in mutability for that motif compared to regular
motifs, minus one. So, <span class="math notranslate nohighlight">\(h^{WRC} &gt; 0\)</span> indicates at hotspot,
<span class="math notranslate nohighlight">\(h^{WRC} &lt; 0\)</span> indicates a coldspot, and <span class="math notranslate nohighlight">\(h^{WRC} = 2\)</span>
indicates a 3x increase in <em>WRC</em> substitution rate. The HLP19 model
by default estimates six motif mutability parameters: four hotspots
(WRC, GYW, WA, and TW) and two coldspots (SYC and GRS).</p>
<p>Substitution models are specified using the <code class="docutils literal notranslate"><span class="pre">-t</span></code> for <span class="math notranslate nohighlight">\(\kappa\)</span>
(transition/transverion rate), <code class="docutils literal notranslate"><span class="pre">--omega</span></code> for <span class="math notranslate nohighlight">\(\omega\)</span>
(nonsynonymous/synonymous mutation rate), and <code class="docutils literal notranslate"><span class="pre">--motifs</span></code> and
<code class="docutils literal notranslate"><span class="pre">--hotness</span></code> for specifying the motif mutability models. The default
for all of these is to estimate shared parameter values across all
lineages, which is also specified by <code class="docutils literal notranslate"><span class="pre">e</span></code>.</p>
<p>Due to default parameter settings, the following two commands are equivalent:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">igphyml</span> <span class="o">--</span><span class="n">repfile</span> <span class="n">ex_lineages</span><span class="o">.</span><span class="n">GY</span><span class="o">.</span><span class="n">tsv</span> <span class="o">-</span><span class="n">m</span> <span class="n">HLP</span> <span class="o">-</span><span class="n">o</span> <span class="n">lr</span>

<span class="n">igphyml</span> <span class="o">--</span><span class="n">repfile</span> <span class="n">ex_lineages</span><span class="o">.</span><span class="n">GY</span><span class="o">.</span><span class="n">tsv</span> <span class="o">-</span><span class="n">m</span> <span class="n">HLP</span> <span class="o">-</span><span class="n">t</span> <span class="n">e</span> <span class="o">--</span><span class="n">omega</span> <span class="n">e</span><span class="p">,</span><span class="n">e</span> \
    <span class="o">--</span><span class="n">motifs</span> <span class="n">WRC_2</span><span class="p">:</span><span class="mi">0</span><span class="p">,</span><span class="n">GYW_0</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span><span class="n">WA_1</span><span class="p">:</span><span class="mi">2</span><span class="p">,</span><span class="n">TW_0</span><span class="p">:</span><span class="mi">3</span><span class="p">,</span><span class="n">SYC_2</span><span class="p">:</span><span class="mi">4</span><span class="p">,</span><span class="n">GRS_0</span><span class="p">:</span><span class="mi">5</span> \
    <span class="o">--</span><span class="n">hotness</span> <span class="n">e</span><span class="p">,</span><span class="n">e</span><span class="p">,</span><span class="n">e</span><span class="p">,</span><span class="n">e</span><span class="p">,</span><span class="n">e</span><span class="p">,</span><span class="n">e</span> <span class="o">-</span><span class="n">o</span> <span class="n">lr</span>
</pre></div>
</div>
<p>In both cases parameter estimates are recorded in
<code class="docutils literal notranslate"><span class="pre">ex_lineages.GY.tsv_igphyml_stats.txt</span></code>. Note that here we use
<code class="docutils literal notranslate"><span class="pre">-o</span> <span class="pre">lr</span></code>, which will keep tree topologies the same and only estimate
branch lengths and substitution parameters. This will keep topologies
the same as the GY94, but will estimate substitution parameters much
more quickly.</p>
<p><strong>Confidence interval estimation</strong></p>
<p>It is possible to estimate 95% confidence intervals for any of these
parameters by adding a ‘c’ to the parameter specification. For example,
to estimate a 95% confidence interval for <span class="math notranslate nohighlight">\(\omega\)</span> in the CDRs
but not the FWRs, use:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">igphyml</span> <span class="o">--</span><span class="n">repfile</span> <span class="n">ex_lineages</span><span class="o">.</span><span class="n">GY</span><span class="o">.</span><span class="n">tsv</span> <span class="o">-</span><span class="n">m</span> <span class="n">HLP</span> <span class="o">-</span><span class="n">o</span> <span class="n">lr</span> <span class="o">--</span><span class="n">omega</span> <span class="n">e</span><span class="p">,</span><span class="n">ce</span>
</pre></div>
</div>
<p>To estimate a 95% confidence interval for <span class="math notranslate nohighlight">\(\omega\)</span> in the FWRs
but not the CDRs, use:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">igphyml</span> <span class="o">--</span><span class="n">repfile</span> <span class="n">ex_lineages</span><span class="o">.</span><span class="n">GY</span><span class="o">.</span><span class="n">tsv</span> <span class="o">-</span><span class="n">m</span> <span class="n">HLP</span> <span class="o">-</span><span class="n">o</span> <span class="n">lr</span> <span class="o">--</span><span class="n">omega</span> <span class="n">ce</span><span class="p">,</span><span class="n">e</span>
</pre></div>
</div>
<p>Any combination of confidence interval specifications can be used
for the above parameter options. For motif mutability, each value
in the <code class="docutils literal notranslate"><span class="pre">--hotness</span></code> option corresponds to the index specified in
the <code class="docutils literal notranslate"><span class="pre">--motifs</span></code> option. To estimate confidence intervals for GYW
mutability, use:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">igphyml</span> <span class="o">--</span><span class="n">repfile</span> <span class="n">ex_lineages</span><span class="o">.</span><span class="n">GY</span><span class="o">.</span><span class="n">tsv</span> <span class="o">-</span><span class="n">m</span> <span class="n">HLP</span> <span class="o">-</span><span class="n">o</span> <span class="n">lr</span> \
    <span class="o">--</span><span class="n">hotness</span> <span class="n">e</span><span class="p">,</span><span class="n">ce</span><span class="p">,</span><span class="n">e</span><span class="p">,</span><span class="n">e</span><span class="p">,</span><span class="n">e</span><span class="p">,</span><span class="n">e</span>
</pre></div>
</div>
<p>which is equivalent to:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">igphyml</span> <span class="o">--</span><span class="n">repfile</span> <span class="n">ex_lineages</span><span class="o">.</span><span class="n">GY</span><span class="o">.</span><span class="n">tsv</span> <span class="o">-</span><span class="n">m</span> <span class="n">HLP</span> <span class="o">-</span><span class="n">o</span> <span class="n">lr</span> \
    <span class="o">--</span><span class="n">motifs</span> <span class="n">WRC_2</span><span class="p">:</span><span class="mi">0</span><span class="p">,</span><span class="n">GYW_0</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span><span class="n">WA_1</span><span class="p">:</span><span class="mi">2</span><span class="p">,</span><span class="n">TW_0</span><span class="p">:</span><span class="mi">3</span><span class="p">,</span><span class="n">SYC_2</span><span class="p">:</span><span class="mi">4</span><span class="p">,</span><span class="n">GRS_0</span><span class="p">:</span><span class="mi">5</span> \
    <span class="o">--</span><span class="n">hotness</span> <span class="n">e</span><span class="p">,</span><span class="n">ce</span><span class="p">,</span><span class="n">e</span><span class="p">,</span><span class="n">e</span><span class="p">,</span><span class="n">e</span><span class="p">,</span><span class="n">e</span>
</pre></div>
</div>
<p>You can also alter constrain motif to have the same mutabilities
by altering the indexes after the ‘:’ in the <code class="docutils literal notranslate"><span class="pre">--motifs</span></code> option.
For example, to estimate 95% confidence intervals on a model in
which WRC/GYW, WA/TW, and SYC/GRS motifs are respectively constrained
to have the same mutabilities, use:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">igphyml</span> <span class="o">--</span><span class="n">repfile</span> <span class="n">ex_lineages</span><span class="o">.</span><span class="n">GY</span><span class="o">.</span><span class="n">tsv</span> <span class="o">-</span><span class="n">m</span> <span class="n">HLP</span> <span class="o">-</span><span class="n">o</span> <span class="n">lr</span> \
    <span class="o">--</span><span class="n">motifs</span> <span class="n">WRC_2</span><span class="p">:</span><span class="mi">0</span><span class="p">,</span><span class="n">GYW_0</span><span class="p">:</span><span class="mi">0</span><span class="p">,</span><span class="n">WA_1</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span><span class="n">TW_0</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span><span class="n">SYC_2</span><span class="p">:</span><span class="mi">2</span><span class="p">,</span><span class="n">GRS_0</span><span class="p">:</span><span class="mi">2</span> \
    <span class="o">--</span><span class="n">hotness</span> <span class="n">ce</span><span class="p">,</span><span class="n">ce</span><span class="p">,</span><span class="n">ce</span>
</pre></div>
</div>
<p>You can find further explanation of the different options in the
commandline help page of BuildTrees, including controlling output
directories and file names.</p>
</div>
<div class="section" id="optimizing-performance">
<h1>Optimizing performance<a class="headerlink" href="#optimizing-performance" title="Permalink to this headline">¶</a></h1>
<p>IgPhyML is a computationally intensive program. There are some ways to
make calculations more practical, however:</p>
<p>Data subsampling: IgPhyML runs slowly with more than a few thousand sequences. You can
subsample your dataset using the <code class="docutils literal notranslate"><span class="pre">--sample</span></code> and <code class="docutils literal notranslate"><span class="pre">--minseq</span></code> options in
BuildTrees.py, which will subsample your dataset to the specified depth and
then remove all clones below the specified size cutoff (see <a class="reference internal" href="#buildtrees-subsampling"><span class="std std-ref">Subsampling
Change-O datasets</span></a>).</p>
<p>Parallelizing computations: It is possible to parallelize likelihood
calculations using the <code class="docutils literal notranslate"><span class="pre">--threads</span></code> option. Currently, calculations
are parallelized by tree, so there is no point in using more threads
than you have lineages in your repertoire file.</p>
<p>GY94 starting topologies: Calculations are much faster under the GY94
model, so it is usually better to <a class="reference internal" href="#building-lineage-trees"><span class="std std-ref">do initial topology
searching under the GY94 model</span></a>,
and then use those trees as starting topologies for HLP19. You can also
keep these topologies the same during HLP19 parameter estimation (<code class="docutils literal notranslate"><span class="pre">-o</span> <span class="pre">lr</span></code>)
for an even greater speedup, though, obviously, this will not result in a
change in topology from GY94.</p>
<p>Enforcing minimum lineage size: Many repertoires often contain huge
numbers of small lineages that can make computations impractical. To
limit the size of lineages being analyzed, specify a cutoff with
<code class="docutils literal notranslate"><span class="pre">--minseq</span></code> when running BuildTrees.py. IgPhyML has a <code class="docutils literal notranslate"><span class="pre">--minseq</span></code> option
with the same functionality, but this option includes the predicted germline
sequence and duplicated sequences in singleton clones. Because of this,
it is recommended to do <code class="docutils literal notranslate"><span class="pre">--minseq</span></code> filtering at the BuildTrees.py step.</p>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../NEWS.html" class="btn btn-neutral float-right" title="Release Notes" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="../install.html" class="btn btn-neutral" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Kenneth B. Hoehn.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../',
            VERSION:'1.0.7',
            LANGUAGE:'None',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../_static/jquery.js"></script>
      <script type="text/javascript" src="../_static/underscore.js"></script>
      <script type="text/javascript" src="../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="../_static/js/theme.js"></script>
  

  <script type="text/javascript">
      jQuery(function () {
          
          SphinxRtdTheme.Navigation.enableSticky();
          
      });
  </script> 

</body>
</html>